gwcs/wcs.py

```diff
@@ -417,6 +417,61 @@ def insert_transform(self, frame, transform, after=False):
             fr, current_transform = self._pipeline[frame_ind]
             self._pipeline[frame_ind] = (fr, transform | current_transform)
 
+    def insert_frame(self, input_frame, transform, output_frame):
+        """
+        Insert a new frame into an existing pipeline. This frame must be
+        anchored to a frame already in the pipeline by a transform. This
+        existing frame is identified solely by its name, although an entire
+        `~gwcs.coordinate_frame.CoordinateFrame` can be passed (e.g., the
+        `input_frame` or `output_frame` attribute). This frame is never
+        modified.
```

Comment:
I think the description of the concept of inserting *a (single) frame* and how it works with this API needs a more clear explanation. For example, if we are inserting *a single frame*, why does the user needs to provide *two* frames? What are the roles of these frames and which one is being inserted: the input one or the output one? The function is called  `insert_frame` but it is not clear neither from the text nor from parameter name which is the frame that is being *inserted*.

I understood how this function is intended to work but it is very confusing and the description should be improved. Some examples that illustrate the purpose and usage of `input_frame` and `output_frame` might help clarify how this function is supposed to be used.

Maybe it would be better to let user provide a tuple of `(Frame, Transform)` and a index where to insert that frame (`int`, `str`, or `Frame` object). I would prefer that this function has a similar usage/behavior as [`insert_transform`](https://github.com/spacetelescope/gwcs/blob/ade33fe478fe5b8784c28fbea1e5179c285670ee/gwcs/wcs.py#L395).

Reply:
This is more complex than `insert_transform` so your suggestion doesn't cover all possibilities. Go to my original post and suppose your `(Frame, Transform)` is `(X, x)`, two parameters which clearly must be specified when calling the method. There are 6 possible final pipelines I could have after inserting this:

`(X, x) (A, a) (B, b) (C, None)`
`(A, a) (X, x) (B, b) (C, None)`
`(A, a) (B, b) (X, x) (C, None)`
`(A, x) (X, a) (B, b) (C, None)`
`(A, a) (B, x) (X, b) (C, None)`
`(A, a) (B, b) (C, x) (X, None)`

So a single frame index as a third parameter can't cover all possibilities. Your suggestion would require a call signature which can be _both_ `insert_frame(Frame, Transform, index/str)` (the first 3 examples, where the `(Frame, Transform)` step goes before the frame with that index/name) and `insert_frame(Transform, Frame, index)` (the last 3 examples, where the transform goes after the frame with that index/name). Python isn't great with variable call signatures and I think this is more confusing that my suggestion. The first call signature here is _exactly_ what I have anyway, while the second is odd because you specify `(x, X, B)` for example, and the output has `x` and `X` _after_ `B`, whereas my suggestion has a call with parameters `(B, x, X)`.

Alternatively, you could fix the parameter order in the call signature to always have `Frame` then `Transform`, and specify the index/name, and then also specify a boolean parameter to say whether you want Frame-then-Transform, or Transform-then-Frame. But now we have 4 parameters instead of 3.

My proposal, where you specify a new frame and new transform and an existing frame, and that existing frame serves as the anchor, remains the most natural looking, IMHO.